pyproject.toml

```diff
@@ -48,6 +48,7 @@ sphinx_rtd_theme = {version="^1.0", optional=true}
 
 [tool.poetry.extras]
 docs = ["sphinx", "sphinx_rtd_theme"]
+grpc = ["grpcio", "protobuf"]
```

Comment:
We need to run tests both with and without extras. `tox.ini` uses `poetry install -v`, which doesn't install extras, but it installs dev-dependencies, which means it still installs `grpcio` and `protobuf`.

Perhaps we should split our dev-dependencies into dependency groups `test` (e.g. pytest) and `codegen` (e.g. grpcio-tools, ...)? https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups

@zhindes @mshafer-NI Do you have any suggestions?

Reply:
We should also add a tox environment called `py39_no_grpc` or something like that. I don't think we need to test each Python version for the no-gRPC case.

Reply:
First draft of dependency groups: keep dev group, convert to Poetry 1.2 syntax, and add codegen group.

```
diff --git a/pyproject.toml b/pyproject.toml
index 69b0999..98c8eea 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -45,24 +45,28 @@ importlib_metadata = {version="^4.2", python="~3.7"}
 # See https://github.com/readthedocs/readthedocs.org/issues/4912.
 Sphinx = {version="^4.3", optional=true}
 sphinx_rtd_theme = {version="^1.0", optional=true}
+grpcio = { version = "^1.49.0", optional = true }
+protobuf = { version = "^4.21", optional = true }

 [tool.poetry.extras]
 docs = ["sphinx", "sphinx_rtd_theme"]
 grpc = ["grpcio", "protobuf"]

-[tool.poetry.dev-dependencies]
-# Test
+[tool.poetry.group.dev.dependencies]
 pytest = ">=7.2"
 pytest-cov = ">=4.0"
 pykka = "^3.0"
 tox = "^3.24"
-click = "^8.1"
-Mako = "^1.2"
 ni-python-styleguide = "^0.4"
 sphinx_rtd_theme = "1.0"
-poetry-grpc-plugin = "^0.1.0"
-grpcio = { version = "^1.49.0", optional = true }
-protobuf = "^3.20"
+mypy = ">=1.0"
+types-protobuf = "^4.21"
+
+[tool.poetry.group.codegen.dependencies]
+click = "^8.1"
+Mako = "^1.2"
+grpcio-tools = "1.49.1"
+mypy-protobuf = ">=3.4"

 [tool.black]
 line-length = 100
@@ -72,5 +76,5 @@ extend_exclude = "docs,generated,src/codegen/metadata,src/codegen/templates"
 extend_exclude = "docs,generated,src/codegen/metadata,src/codegen/templates,src/handwritten"

 [build-system]
-requires = ["poetry>=1.1.12"]
+requires = ["poetry>=1.2"]
 build-backend = "poetry.masonry.api"
\ No newline at end of file
```

I verified that `poetry1.2 install --without codegen` does not install grpcio, protobuf, click, etc., and that `poetry1.2 install --without codegen --extras grpc` installs grpcio and protobuf.

Reply:
As suggested, I have added the dependency groups and updated the poetry version. Do we need to create a tox environment as part of this PR or can we take that up while updating the tests?

Reply:
You can add the tox environment when updating the tests.